collectors/channels_collector.go

```diff
@@ -317,6 +340,150 @@ func (c *ChannelsCollector) Collect(ch chan<- prometheus.Metric) {
 			)
 		}
 	}
+
+	// Get all remote policies
+	remotePolicies, err := c.getRemotePolicies(getInfoResp.IdentityPubkey)
+	if err != nil {
+		channelLogger.Error(err)
+		return
+	}
+
+	// Export the inbound fee metric for a series of amounts.
+	var feeAmount btcutil.Amount = 100000
+	for {
+		inboundFee := getInboundFee(feeAmount, remotePolicies, remoteBalances)
+		if inboundFee == nil {
+			break
+		}
+		ch <- prometheus.MustNewConstMetric(
+			c.inboundFee, prometheus.GaugeValue,
+			float64(*inboundFee)/float64(feeAmount),
+			feeAmount.String(),
+		)
+
+		feeAmount *= 2
+	}
+}
+
+// getInboundFee calculates to forward fee for a specific amount charged by the
+// last hop before this node.
+func getInboundFee(amt btcutil.Amount, remotePolicies map[uint64]*lnrpc.RoutingPolicy,
+	remoteBalances map[uint64]btcutil.Amount) *btcutil.Amount {
+
+	var fee btcutil.Amount
+
+	// Copy the remote balances so they can be decreased as we find shards.
+	remainingBalances := make(map[uint64]btcutil.Amount)
+	for ch, balance := range remoteBalances {
+		remainingBalances[ch] = balance
+	}
+
+	// Assume a perfect mpp splitting algorithm that knows exactly how much
+	// can be sent through each channel. This is a simplification, because
+	// in reality senders need to trial and error to find a shard amount
+	// that works.
+	amountRemaining := amt
+	for amountRemaining > 0 {
+		var (
+			bestChan        uint64
+			bestSpecificFee float64
+			bestAmount      btcutil.Amount
+			bestFee         btcutil.Amount
+		)
+
+		// Find the best channel to send the amount or a part of the
+		// amount.
+		for ch, balance := range remainingBalances {
+			// Skip channels without remote balance.
+			if balance == 0 {
+				continue
+			}
+
+			policy, ok := remotePolicies[ch]
+			if !ok {
+				continue
+			}
+
+			// Cap at the maximum receive amount for this channel.
+			amountToSend := amountRemaining
+			if amountToSend > balance {
+				amountToSend = balance
+			}
+
+			// Calculate fee for this amount to send.
+			fee := btcutil.Amount(
+				policy.FeeBaseMsat/1000 +
+					int64(amountToSend)*policy.FeeRateMilliMsat/1000000,
+			)
+
+			// Calculate the specific fee for this amount, being the
+			// fee per sat sent.
+			specificFee := float64(fee) / float64(amountToSend)
+
+			// Select the best channel for this shard based on the
+			// lowest specific fee.
+			if bestChan == 0 || bestSpecificFee > specificFee {
```

Comment:
>If a channel has low balance then it could also have the best specific fee right?

Yes, the fee rate and base fee could be 0 even, so that makes the specific fee zero too.